### PR TITLE
Activation selection

### PR DESCRIPTION
**Pull request by jumelet**
Makes `ActivationReader` indexable, either by position or sentence key. Indexing can be done by `int`, `slice` or a list/array of tuples.

### CODE REVIEW

File: rnnalyse/extractors/base_extractor.py [hunk 1]
```diff
@@ -86,15 +86,16 @@ def extract(self,
             be extracted or not.
         """
         start_t = prev_t = time()
-        num_extracted = n_sens = 0
+        n_extracted = n_sens = 0
```
Comment (Kaleidophon) [inline rnnalyse/extractors/base_extractor.py]: Is it necessary to initialize `n_sens` here? 🤔

File: rnnalyse/models/language_model.py [hunk 2]
```diff
@@ -3,23 +3,27 @@
...
 
-from ..typedefs.models import FullActivationDict
+from ..typedefs.activations import FullActivationDict
 
 
 class LanguageModel(nn.Module):
     """ Abstract class for LM with intermediate activations """
-    def __init__(self) -> None:
-        super(LanguageModel, self).__init__()
+    def __init__(self,
```
Comment (Kaleidophon) [inline rnnalyse/models/language_model.py]: Did you opt for the underscore names here because this class is supposed to be used as an abstract class?

File: rnnalyse/activations/activation_reader.py [hunk 3]
```diff
@@ -22,34 +23,154 @@ class ActivationReader:
...
+        self.label_path = label_path
+
+        self._activations = None
+        self._labels: Optional[np.ndarray] = None
+        self._data_len: int = -1
+        self._activation_ranges: Optional[ActivationRanges] = None
+
+        self.activations = None
+
+    # TODO: Add examples here or somewhere else
```
Comment (Kaleidophon) [inline rnnalyse/activations/activation_reader.py]: I think indeed putting some examples into the docstring of this function or the class would be very helpful

File: rnnalyse/activations/activation_reader.py [hunk 4]
```diff
@@ -22,34 +23,154 @@ class ActivationReader:
...
+    def __getitem__(self, key: ActivationKey) -> np.ndarray:
+        """ Provides indexing of activations, indexed by sentence
+        position or key, or indexing the activations itself. Indexing
+        based on sentence returns all activations belonging to that
+        sentence.
+
+        Indexing by position ('pos') refers to the order of extraction,
+        selecting the first sentence ([0]) will return all activations
+        of the sentence that was extracted firstly.
```
Comment (Kaleidophon) [inline rnnalyse/activations/activation_reader.py]: *first

File: rnnalyse/activations/activation_reader.py [hunk 5]
```diff
@@ -22,34 +23,154 @@ class ActivationReader:
...
+
+        The index, indexing type and activation name can be provided as:
+          [index]
+        | [index, indextype]         | [index, a_name]
+        | [index, a_name, indextype] | [index, indextype, a_name]
+        With indextype either 'pos', 'key' or 'all', and activation name
+        a (layer, name) tuple.
+
+        If activationname is not provided it should have been set
+        beforehand like `reader.activations = activationname`.
```
Comment (Kaleidophon) [inline rnnalyse/activations/activation_reader.py]: So if I understand this correctly, if `reader.activations` is set even only using `[index]` will return the activations corresponding to `activationname`?

File: rnnalyse/activations/activation_reader.py [hunk 6]
```diff
@@ -22,34 +23,154 @@ class ActivationReader:
...
+        if indextype == 'all':
+            return self.activations[index]
+        elif indextype == 'key':
+            ranges = self._create_range_from_key(index)
+        else:
+            ranges = np.array(list(self.activation_ranges.values()))[index]
+        inds = self._create_indices_from_range(ranges)
+        return self.activations[inds]
+
+    def _parse_key(self, key: ActivationKey) -> Tuple[ActivationIndex, str]:
```
Comment (Kaleidophon) [inline rnnalyse/activations/activation_reader.py]: I think some short comments would help here to understand which kind of key lands in which if/else block

File: rnnalyse/extractors/base_extractor.py [hunk 7]
```diff
@@ -110,19 +111,25 @@ def extract(self,
                         all_activations[name].append(sen_activations[name])
 
                 extracted_labels.extend(sen_extracted_labels)
-                num_extracted += len(sen_extracted_labels)
+                activation_ranges[sen_id] = (n_extracted, n_extracted+len(sen_extracted_labels))
```
Comment (Kaleidophon) [inline rnnalyse/extractors/base_extractor.py]: So `activation_ranges` is  a dict mapping from sentence_id to the positions that were extracted? I suppose this comes in handy somewhere when using the new indexing for `ActivationsReader`?

File: rnnalyse/extractors/base_extractor.py (hunk 1 above)
Comment (jumelet) [reply rnnalyse/extractors/base_extractor.py]: So in practice it isn't, but to be completely type-correct you should initialize a variable that is used in a for loop if you want to use it afterwards.

So omitting `n_sens` here wouldn't necessarily break anything, only when `self.corpus` is empty.

File: rnnalyse/models/language_model.py (hunk 2 above)
Comment (jumelet) [reply rnnalyse/models/language_model.py]: Yes, although I can imagine this setup is not perfect yet. Do you have more experience with setting up abstract classes? I'm not sure how something like `abc` can be used here if we still want `LanguageModel` to inherit from `nn.Module`.

Comment (jumelet) [reply rnnalyse/models/language_model.py]: Let me know what you think of my latest addition to this class, not super sure if multiple inheritance is the way to go here but my compiler isn't complaining :p

File: rnnalyse/activations/activation_reader.py (hunk 5 above)
Comment (jumelet) [reply rnnalyse/activations/activation_reader.py]: Exactly. My idea for this is that this allows you to simply set/read the activations only once, and have all your subsequent indexing being applied to those activations.

File: rnnalyse/extractors/base_extractor.py (hunk 7 above)
Comment (jumelet) [reply rnnalyse/extractors/base_extractor.py]: Yes indeed, this makes it possible to retrieve the activations of a specific sentence. Because of the selection_func (which is a fantastic addition btw!), the ranges can no longer be based on the sentence lengths.

I can imagine that in the future we want to write more relevant info to file besides these ranges, but for now I could only come up with the ranges being useful. Maybe dumping the actual extracted indices of a sentence is nice as well of course, because right now you only know which activations belong to some sentence, but not to the individual tokens themselves. I'll add it to the todo's :p

File: rnnalyse/models/language_model.py (hunk 2 above)
Comment (dieuwkehupkes) [reply rnnalyse/models/language_model.py]: I think this works,  but is it even necessary to inherit from ABC. The class is already abstract because it has some abstract methods that need to be implemented. Maybe just inheriting from LanguageModel is enough?

File: rnnalyse/activations/activation_reader.py [hunk 8]
```diff
@@ -22,34 +23,166 @@ class ActivationReader:
...
     ----------
     activations_dir : str
-    labels : np.ndarray
-        Numpy array containing the extracted labels
-    data_len : int
-        Number of extracted activations
+    label_path : str
+    activations : Optional[np.ndarray]
+        Numpy array of activations that are currently read into ram.
+    _labels : Optional[np.ndarray]
```
Comment (dieuwkehupkes) [inline rnnalyse/activations/activation_reader.py]: I thought that these were the underscores that @Kaleidophon was referring to before, but apparently I misunderstood. Why are these attributes underscored? Is it a convention to denote they are private?

Comment (jumelet) [reply rnnalyse/activations/activation_reader.py]: Oh wait, I think I misunderstood @Kaleidophon 's remark as well. It is indeed a convention to denote variables that are intended for internal use, this blog explains it well: https://dbader.org/blog/meaning-of-underscores-in-python

I use it there because `self.labels` now refers to the `@property` which returns the actual array that is stored in `._labels`.

However, in `LanguageModel` I mainly used it because putting an `_` in front of it stopped my checker from complaining about singleton variables. 

Thinking about this, it makes much more sense of course to use `**kwargs` in the `LanguageModel` init, I'll update it.

Comment (jumelet) [reply rnnalyse/activations/activation_reader.py]: Sorry for all the rambling, but the only thing I'm uncertain about right now is that `LanguageModel` doesn't tell you at all which arguments it's expecting from its subclasses. Maybe `*args` isn't too perfect after all, putting the explicit arguments there did make that more clear.

What do you think?

Comment (jumelet) [reply rnnalyse/activations/activation_reader.py]: Adding `@abstractmethod` to the init seems to stop pylint from complaining, and also makes the expected arguments more explicit. I think this is the way to go then.

File: rnnalyse/models/language_model.py (hunk 2 above)
Comment (Kaleidophon) [reply rnnalyse/models/language_model.py]: I think it looks a bit strange because you use underscore variables and names for protected attributes and class methods. I would just inherit from ABC (abstract base class) from the `abc` package as well, in this class you also get an error if you try to instantiate an object from that class.

File: rnnalyse/extractors/base_extractor.py (hunk 7 above)
Comment (Kaleidophon) [reply rnnalyse/extractors/base_extractor.py]: I see, then this addition makes a lot of sense! Thanks!

File: rnnalyse/activations/activation_reader.py (hunk 8 above)
Comment (Kaleidophon) [reply rnnalyse/activations/activation_reader.py]: I think if you're using `@property` it is fine to use something like `_labels`. In the docstring I would just refer to `labels` because the `labels` function decorated with `@property`  essentially functions like an attribute.

Comment (Kaleidophon) [reply rnnalyse/activations/activation_reader.py]: And why was Pylint complaining about init? Using `@abstractmethod` on init doesn't seem to make a lot of sense to me

Comment (jumelet) [reply rnnalyse/activations/activation_reader.py]: I guess it did because a non-abstractmethod is expected to do something with its arguments. I got the idea of making init an abstractmethod from this SO discussion: https://stackoverflow.com/questions/44800659/python-abstract-class-init

So to couple it back to why we're doing all this: we want the `LanguageModel` to be an abstract class which ensures its subclasses adhere to the same init structure (model/vocab/module/device). To ensure this we don't want to use generic `*args` in the init, as this would allow a subclass to input any kind of stuff it wants there.

Furthermore, we don't want to simply set these arguments as class attributes in `LanguageModel`, as for instance the `model` argument is only the path to the model and it is up to the subclass to load in the model itself. On the other hand we could of course add an abstractmethod to `LanguageModel` that should take care of loading in the model/vocab/module stuff. This might make the usecase a bit more explicit.

Comment (Kaleidophon) [reply rnnalyse/activations/activation_reader.py]: Okay I just fiddled around a bit and I have to say that I was wrong and just putting `ABC` as a superclass doesn't prevent you from instantiating it, you have to indeed add an `@abstractmethod` to init. However, if you do override the (default super-)constructor you can basically pass an abritrary signature and it still works (your IDE might warn you though, I guess that is what we're aiming for?).

The example I is [here](https://trinket.io/python3/7fd5b76925).

Comment (jumelet) [reply rnnalyse/activations/activation_reader.py]: Yeah nice, it starts to make sense now.

If you'd override the init constructor with a different signature pylint will complain, so that's nice. It is already complaining right now about the `forward` overriding the `forward` of `nn.module`, which might be kinda related to the point you raised about training a new language model and all that. Food for thought, I can imagine we want to change it again somewhere in the future
